Mark as read functionality

AmaranthineCodices opened the pull request:
This adds:
- A database migration system for the backend (see `migration.rs`)
  Migrations are simple - we store a persistent schema version in SQLite's `user_version` pragma. This version is an index into a `Vec` of migration functions, which upgrade the database schema from the previous version to their current version. In the future, how we organize these probably needs to be revisited - right now it's stuck in one file.
- Database schema version 1 (version 0 is current `master`) adds an `isRead` column to the `feedEntries` table. This column is an integer; since SQLite does not have a boolean type, an integer is used as a replacement. The frontend is not aware of this.
- A `MarkRead` request type has been added, and a corresponding action has been added to the frontend's Redux store.
- Feed entries can be marked as "read" by clicking on them. This will currently gray out the background. This is persistent and stored.

TODO before merging:
- [ ] ~~Try to simplify the mess that is the frontend implementation of marking posts as read.~~
- [x] Maybe generalize the requests to `SetRead` or something similar, so we can mark a post as unread?

TODO after merging, in subsequent PR:
- Add an option to filter out unread posts from the frontend

AmaranthineCodices committed:
database migration system 🎉

AmaranthineCodices committed:
serve row_id to clients

AmaranthineCodices committed:
set up server-side marking as read

AmaranthineCodices committed:
MarkAsRead -> MarkRead

AmaranthineCodices committed:
export is_read to clients

AmaranthineCodices committed:
finish migrating MarkAsRead -> MarkRead, apparently

AmaranthineCodices committed:
specify isRead value

AmaranthineCodices committed:
I did a thing (it's awful)

tiffany352 commented on backend/src/main.rs:
```diff
@@ -49,6 +50,11 @@ fn main() {
 
     create_db();
```
Why don't we make create_db() into a migration?

AmaranthineCodices replied:
...because I wasn't thinking about it right and made it more complicated than it needed to be :p

Nothing stops this from being a migration.

tiffany352 commented on backend/src/protocol/request.rs:
```diff
@@ -14,12 +14,18 @@ pub struct AddFeed {
 }
 
+#[derive(Deserialize, Debug)]
+pub struct MarkRead {
```
Yeah this should definitely be a SetReadStatus with a bool for whether to mark read. It might make sense to make it a batch API as well, not sure.

AmaranthineCodices replied:
In the context of "mark a specific entry as read" a batch API doesn't make much sense, but now that I've had time to think about it, a batch API could be useful for a feature like "mark this entire feed as read".

tiffany352 commented on frontend/src/components/FeedEntryList.tsx:
```diff
@@ -1,17 +1,32 @@
 class FeedEntryList extends React.Component<FeedEntryListProps> {
+  constructor(props: FeedEntryListProps) {
+    super(props);
+    this.markRead = this.markRead.bind(this);
```
Should we be using arrow notation for these? Then we don't need the constructor.

AmaranthineCodices replied:
We should, now that I see it! I didn't know you could use arrow notation in classes like that :o

AmaranthineCodices committed:
move create_db into a migration

AmaranthineCodices committed:
batched, more general SetRead request instead of just MarkRead

AmaranthineCodices committed:
use transaction

AmaranthineCodices committed:
use prepare_cached to avoid recompiling statements

AmaranthineCodices committed:
line frontend up with backend

- markRead -> setRead
- entries can be "unread" by clicking them again
- setRead operates on an array of entries, not just one entry

tiffany352 commented on backend/src/requests/set_read.rs:
```diff
@@ -0,0 +1,27 @@
+use super::Error;
+
+pub fn set_read(request: SetRead, conn: &mut Connection) -> Result<Response, Error> {
+	let tx = conn.transaction()?;
```
I think the convention for rust is 4 spaces, but this file uses tabs?

AmaranthineCodices committed:
fix indentation

AmaranthineCodices committed:
add editorconfig
